ee/auth/saml/router.js

```diff
@@ -83,8 +83,30 @@ router.post(
       throw new Error('Missing one or more SAML attributes');
     }
 
-    const params = [authUid, authName, authUin, 'SAML'];
-    const userRes = await sqldb.callAsync('users_select_or_insert', params);
+    let userRes;
+    try {
+      userRes = await sqldb.callAsync('users_select_or_insert', [
+        authUid,
+        authName,
+        authUin,
+        'SAML',
+      ]);
+    } catch (err) {
+      // Check if this is because the provider is not allowed.
+      if (err.message?.includes('authentication provider is not allowed for institution')) {
+        // Parse institution ID from error message.
+        const institutionIdMatch = err.message.match(/\((\d+)\)$/);
+        if (institutionIdMatch) {
+          const institutionId = institutionIdMatch[1];
+          res.redirect(`/pl/auth/institution/${institutionId}/not_allowed`);
+          return;
+        }
+      }
```

Comment:
I copy/pasted this block of code around, which is probably no worse than all the other auth code we've duplicated in 4 places?

Grepping through the error message to tell what it is and which institution is in play is kinda gross too 🙃

Reply:
Yeah, this is indeed rather gross.

How about we return a `result` from `users_select_or_insert` which is either `"success"` or `"invalid_provider"` and also return the `institution_id` in all cases? (and still raise exceptions if something else weird goes wrong)

From what I can see we don't have all that many call sites for `users_select_or_insert` so it wouldn't be too bad to check a `result` at each of them?

That being said, I won't block on this if you prefer to parse the `institution_id` out of the error message.

Reply:
SGTM! I'll make the change.

Reply:
Done in 8cbdf8c. The code looks much cleaner now!

Reply:
Nice!

Reply:
Do we need to modify this call in any way? https://github.com/PrairieLearn/PrairieLearn/blob/7a5ab555f9ea094af519f7ec293839ca22773aaf/middlewares/authn.js#L105

I'm actually not sure we want to be calling `users_select_or_insert` here, rather than just a `users_select`. I don't think that `insert` should ever happen from this call site?

Maybe it's fine to call `users_select_or_insert ` here and ignore the invalid-institution case because it can't happen here?